ext/standard/math.c

```diff
@@ -228,6 +228,32 @@ PHP_FUNCTION(abs)
 }
 /* }}} */
 
+/* {{{ Return the given value if in range of min and max */
+PHP_FUNCTION(clamp)
```

Comment:
how php engine knowns this function is deterministic and time stable, eg. can be evaluated at compile time?

Reply:
This is handled in OpCache, and can be done later.

Reply:
Yeah, my knowledge of OpCache is absolute zero, so to be honest I sadly cannot answer that question. 🤷‍♂️

Reply:
FWIW this is just a matter of adding the function to https://github.com/php/php-src/blob/64525b6c1a370e733ea246cc055b4bb7f76a07e0/Zend/Optimizer/sccp.c#L799. Though I'm not sure how practically useful that would really be. Notably we don't include min or max there either.

Reply:
Wow, I belive fixing this for every function can increase the performance by at least 5%. Would be great if we can add some flag for it to the code directly (and to the stubs which can be then used also by a static analysers).

`min`/`max` and other simple functions are used often in loop conditions and evaluating them only once should be definitely standard. For some reasons, there is `pow` present, but `sqrt` not, `dirname` is present, but `basename` not. Flag directly at the function declaration should fix this unmainatained/uncomplete list.

Reply:
This list is only relevant to compile-time evaluation, not for loop invariant code motion. You need to actually have something like `min(1.0, 2.0)` in source code (or at least, we need to be able to determine the argument values through constant propagation). I doubt this is a common pattern. This list is mostly based on functions where compile-time evaluation has been found useful in some cases -- e.g. `dirname(__FILE__)` is a common pattern in older code.

> but basename not.

This is because basename is locale-dependent.

Reply:
Something like `str_repeat('test', min(10, Class::MAX_ITERS))` is quite common